### PR TITLE
[Example] Implement deepwalk by dgl and pytorch

### PR DESCRIPTION
**Pull request by ShawXh**
## Description

Deepwalk is implemented by functions of dgl with pytorch as backend. It has the following features:
* It can be run on CPU, GPU, and mixed CPU with multi-GPU
* It provides both SGD and Adam for gradient descent
* The performance is competitive compared with other implementations, provided in README
* It adopts some tricks on negative sampling to reduce data transfer between CPU and GPU
* It updates node embedding directly without 'forward' and 'backward'
* It does sigmoid by a lookup table, avoiding computing the accurate value

Performance on YouTube (1M nodes).

| Implementation | Macro-F1 (%) <br> 1% &emsp;&emsp; 3% &emsp;&emsp; 5% &emsp;&emsp; 7% &emsp;&emsp; 9% | Micro-F1 (%) <br> 1% &emsp;&emsp; 3% &emsp;&emsp; 5% &emsp;&emsp; 7% &emsp;&emsp; 9% |
|----|----|----|
| gensim.word2vec(hs) | 28.73 &emsp; 32.51 &emsp; 33.67 &emsp; 34.28 &emsp; 34.79 | 35.73 &emsp; 38.34 &emsp; 39.37 &emsp; 40.08 &emsp; 40.77 | 
| gensim.word2vec(ns) | 28.18 &emsp; 32.25 &emsp; 33.56 &emsp; 34.60 &emsp; 35.22 | 35.35 &emsp; 37.69 &emsp; 38.08 &emsp; 40.24 &emsp; 41.09 | 
|        Ours         | 24.58 &emsp; 31.23 &emsp; 33.97 &emsp; 35.41 &emsp; 36.48 | 38.93 &emsp; 43.17 &emsp; 44.73 &emsp; 45.42 &emsp; 45.92 | 

The comparison between running time is shown as below, where the numbers in the brackets are time used on random-walk. Ours is run on mixed CPU and 4GPUs.

| Implementation | gensim.word2vec(hs) | gensim.word2vec(ns) | Ours |
|----|----|----|----|
| Time (s) |    27119.6(1759.8)  |   10580.3(1704.3)   | 428.89 |

Speeding-up with mixed CPU & multi-GPU. The used parameters are the same as above.
|  #GPUs   |   1   |   2   |   4   |
|----------|-------|-------|-------|
| Time (s) |1419.64| 952.04|428.89 |

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

## Changes

It is an initial version.

### CODE REVIEW

File: examples/pytorch/deepwalk/deepwalk.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    elif args.mix:
+        print("Mix CPU with %d GPU" % args.num_procs)
+        if args.num_procs == 1:
+            trainer.emb_model.set_device(0)
+    else:
+        print("Run in %d CPU process" % args.num_procs)
+
+def fast_train_mp(trainer, args):
+    """ multi-cpu or mix multi-gpu """
+    set_device(trainer, args)
```
Comment (classicsong) [inline examples/pytorch/deepwalk/deepwalk.py:97]: You can calculate device here, and use this device in this func.

File: examples/pytorch/deepwalk/deepwalk.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    trainer.share_memory()
+    random.shuffle(trainer.dataset.walks)
```
Comment (classicsong) [inline examples/pytorch/deepwalk/deepwalk.py:99]: You donot shuffle walks, but use:
```
seeds = th.torch.randperm(number_of_nodes)
```

File: examples/pytorch/deepwalk/deepwalk.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    trainer.emb_model.set_device(gpu_id)
+    torch.set_num_threads(args.num_threads)
+
+    start = time.time()
+    with torch.no_grad():
+        i = 0
+        max_i = args.iterations * num_batches
+        
+        while True:
+            lr = args.lr * (max_i - i) / max_i
```
Comment (classicsong) [inline examples/pytorch/deepwalk/deepwalk.py:134]: Add docstr here

File: examples/pytorch/deepwalk/model.py [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+        if self.adam:
+            lr = self.lr
+
+        # [batch_size, walk_length]
+        nodes = torch.stack(batch_walks)
+        if self.only_gpu:
+            nodes = nodes.to(self.device)
+            if neg_nodes is not None:
+                neg_nodes = neg_nodes.to(self.device)
+        emb_u = self.u_embeddings.weight[nodes].view(-1, self.emb_dimension).to(self.device)
```
Comment (classicsong) [inline examples/pytorch/deepwalk/model.py:289]: you can use

emb_u = self.u_embeddings(nodes).view(-1, self.emb_dimension).to(self.device)
emb_v = self.v_embeddings(nodes).view(-1, self.emb_dimension).to(self.device)

**Comment by classicsong** [comment]
Thank you for your contribution.
